File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/API/Advancement/AdvancementMessage.java
package jp.aoichaan0513.A_TosoGame_Live.API.Advancement;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.advancement.Advancement;
import org.bukkit.advancement.AdvancementProgress;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.Collection;
import java.util.Collections;

public class AdvancementMessage {

    private final NamespacedKey id;
    private final String icon;
    private final String title;
    private final String description;

    public AdvancementMessage(String id, String title, String description, Material icon) {
        this(new NamespacedKey(Main.getInstance(), id), title, description, icon);
    }

    public AdvancementMessage(NamespacedKey id, String title, String description, Material icon) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.icon = NamespacedKey.minecraft(icon.name().toLowerCase()).toString();
    }

    public void showTo(Player player) {
        showTo(Collections.singletonList(player));
    }

    public void showTo(Collection<? extends Player> players) {
        add();
        grant(players);
        new BukkitRunnable() {

            @Override
            public void run() {
                revoke(players);
            }
        }.runTaskLater(Main.getInstance(), 20);
    }

    private void add() {
        try {
            Bukkit.getUnsafe().loadAdvancement(id, getJson());
            Bukkit.getLogger().info("Advancement " + id + " saved");
        } catch (IllegalArgumentException e) {
            Bukkit.getLogger().info("Error while saving, Advancement " + id + " seems to already exist");
        }
    }

    private void remove() {
        Bukkit.getUnsafe().removeAdvancement(id);
    }

    private void grant(Collection<? extends Player> players) {
        Advancement advancement = Bukkit.getAdvancement(id);
        AdvancementProgress progress;
        for (Player player : players) {

            progress = player.getAdvancementProgress(advancement);
            if (!progress.isDone())
                for (String criteria : progress.getRemainingCriteria())
                    progress.awardCriteria(criteria);
        }
    }

    private void revoke(Collection<? extends Player> players) {
        Advancement advancement = Bukkit.getAdvancement(id);
        AdvancementProgress progress;
        for (Player player : players) {

            progress = player.getAdvancementProgress(advancement);
            if (progress.isDone())
                for (String criteria : progress.getAwardedCriteria())
                    progress.revokeCriteria(criteria);
        }
    }

    public String getJson() {
        JsonObject iconObject = new JsonObject();
        iconObject.addProperty("item", this.icon);

        JsonObject displayObject = new JsonObject();
        displayObject.add("icon", iconObject);
        displayObject.addProperty("title", this.title);
        displayObject.addProperty("description", this.description);
        displayObject.addProperty("background", "minecraft:textures/gui/advancements/backgrounds/adventure.png");
        displayObject.addProperty("frame", "task");
        displayObject.addProperty("announce_to_chat", false);
        displayObject.addProperty("show_toast", true);
        displayObject.addProperty("hidden", true);

        JsonObject triggerObject = new JsonObject();
        triggerObject.addProperty("trigger", "minecraft:impossible");

        JsonObject criteriaObject = new JsonObject();
        criteriaObject.add("impossible", triggerObject);

        JsonObject jsonObject = new JsonObject();

        jsonObject.add("criteria", criteriaObject);
        jsonObject.add("display", displayObject);

        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        return gson.toJson(jsonObject);

    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/API/Advancement/AdvancementMessage.kt
package jp.aoichaan0513.A_TosoGame_Live.API.Advancement

import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import jp.aoichaan0513.A_TosoGame_Live.Main
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.advancement.AdvancementProgress
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable

class AdvancementMessage(private val id: NamespacedKey, private val title: String, private val description: String, icon: Material) {
    private val icon: String

    constructor(id: String, title: String, description: String, icon: Material) : this(NamespacedKey(Main.pluginInstance, id), title, description, icon)

    fun showTo(player: Player) {
        showTo(listOf(player))
    }

    fun showTo(players: Collection<Player>) {
        add()
        grant(players)
        object : BukkitRunnable() {
            override fun run() {
                revoke(players)
            }
        }.runTaskLater(Main.pluginInstance, 20)
    }

    private fun add() {
        try {
            Bukkit.getUnsafe().loadAdvancement(id, json)
            Bukkit.getLogger().info("Advancement $id saved")
        } catch (e: IllegalArgumentException) {
            Bukkit.getLogger().info("Error while saving, Advancement $id seems to already exist")
        }
    }

    private fun remove() {
        Bukkit.getUnsafe().removeAdvancement(id)
    }

    private fun grant(players: Collection<Player>) {
        val advancement = Bukkit.getAdvancement(id)
        var progress: AdvancementProgress
        for (player in players) {
            progress = player.getAdvancementProgress(advancement!!)
            if (!progress.isDone)
                for (criteria in progress.remainingCriteria)
                    progress.awardCriteria(criteria)
        }
    }

    private fun revoke(players: Collection<Player>) {
        val advancement = Bukkit.getAdvancement(id)
        var progress: AdvancementProgress
        for (player in players) {
            progress = player.getAdvancementProgress(advancement!!)
            if (progress.isDone)
                for (criteria in progress.awardedCriteria)
                    progress.revokeCriteria(criteria)
        }
    }

    val json: String
        get() {
            val iconObject = JsonObject()
            iconObject.addProperty("item", icon)
            val displayObject = JsonObject()
            displayObject.add("icon", iconObject)
            displayObject.addProperty("title", title)
            displayObject.addProperty("description", description)
            displayObject.addProperty("background", "minecraft:textures/gui/advancements/backgrounds/adventure.png")
            displayObject.addProperty("frame", "task")
            displayObject.addProperty("announce_to_chat", false)
            displayObject.addProperty("show_toast", true)
            displayObject.addProperty("hidden", true)
            val triggerObject = JsonObject()
            triggerObject.addProperty("trigger", "minecraft:impossible")
            val criteriaObject = JsonObject()
            criteriaObject.add("impossible", triggerObject)
            val jsonObject = JsonObject()
            jsonObject.add("criteria", criteriaObject)
            jsonObject.add("display", displayObject)
            val gson = GsonBuilder().setPrettyPrinting().create()
            return gson.toJson(jsonObject)
        }

    init {
        this.icon = NamespacedKey.minecraft(icon.name.toLowerCase()).toString()
    }
}